### PR TITLE
Allow filtering in value relation widget when allowing multiple selections

### LINKED ISSUE #50371: Multiple selections and completer does not work together in Value Relation
### What is the bug or the crash?

If I activate both "Allow multiple selections" and "Use completer", only multiple selections is possible/available. It is not possible start entering a value to get a list with the option(s).
image

Nowhere to start entering a value (a species name in this case):
image

(If I don't activate "Allow multiple selections", the completer appears and I can start entering a name:
image)

I may combine the mentioned options with "Allow NULL values" and "Order by value", but doesn't the use of check boxes indicate that you may check whatever options you want, e.g. all four of them?

### Steps to reproduce the issue

Having a list (CSV-file) and a point-layer that may be joined through value relation.
1. Go to Layer Properties - Attributes Form - Autogenerate
2. Click on the the field and set widget type to "Value Relation"
3. Check both "Allow multiple selections" and "Use completer".
4. Set point layer in edit mode and add a point. 
5. In the Features Attributes window, the completer will be missing.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.26.3-Buenos Aires | QGIS code revision | 65e4edfdada
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.5.1
PROJ version | 9.1.0
EPSG Registry database version | v10.074 (2022-08-01)
GEOS version | 3.10.3-CAPI-1.16.1
SQLite version | 3.38.1
PDAL version | 2.4.3
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
autoSaver | 2.6
DataPlotly | 3.9.2
get_nib-main | 0.7
GroupStats | 2.2.6
InvisibleLayersAndGroups | 2.2
LAStools | 1.4
legend_data_filter | 0.1
plugin_reloader | 0.9.1
Projestions | 1.0.1
qfieldsync | v4.2.0
Smart_Map | 1.2
valuetool | 3.0.15
db_manager | 0.1.20
grassprovider | 2.12.99
processing | 2.12.99

</body></html>QGIS version
3.26.3-Buenos Aires
QGIS code revision
[65e4edfdada](https://github.com/qgis/QGIS/commit/65e4edfdada)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.5.1
PROJ version
9.1.0
EPSG Registry database version
v10.074 (2022-08-01)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.38.1
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
autoSaver
2.6
DataPlotly
3.9.2
get_nib-main
0.7
GroupStats
2.2.6
InvisibleLayersAndGroups
2.2
LAStools
1.4
legend_data_filter
0.1
plugin_reloader
0.9.1
Projestions
1.0.1
qfieldsync
v4.2.0
Smart_Map
1.2
valuetool
3.0.15
db_manager
0.1.20
grassprovider
2.12.99
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

I was hoping for the same behavior as in Input from Mergin Maps [(https://merginmaps.com/docs/layer/settingup_forms/#value-relation)](https://merginmaps.com/docs/layer/settingup_forms/#value-relation) based on QGIS core library):
image

### PR DESCRIPTION
**Pull request by uclaros**
## Description
When using the Value Relation widget on a form and having _Allow multiple selections_ enabled, the items were added to a `QTableWidget` with the specified number of columns.
This PR uses a new widget instead that also offers a `QgsFilterLineEdit` allowing filtering the displayed table's items.
The line edit widget is only displayed when both  _Allow multiple selections_ and _Use completer_ are enabled.
When _Use completer_ is disabled, the functionality should be unchanged.

Here is a sample with two fields, one with 5 columns and completer, one with 3 columns and no completer, using lots of bird names as values:

https://user-images.githubusercontent.com/11358178/207796644-c50d713c-2786-43a0-8949-0fa388281783.mp4

Fixes #50371
Fixes unreported bug: after stopping editing on the attribute table while still editing a cell, the next cell appears editable, related to #8524

### CODE REVIEW

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h [hunk 1]
```diff
@@ -16,17 +16,98 @@
...
+
+    /**
+     * Sets all items to be partially checked
+     */
+    void setIndeterminateState();
+
+    /**
+     * Set all table items to \a enabled.
+     */
+    void setEnabledTable( const bool enabled );
```
Comment (troopa81) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:78]: nitpick: why not just setEnabled ?

**Review by troopa81 — COMMENTED** [review]
Nice feature :+1: 

Have you considered using QAbstractItemModel instead of a QTableWidget since all the model logic is within mCache ? It will avoid allocating QTableWidget every time filter string changes, no?

Besides that, is it possible to make the space between lines constant so it doesn't grow when we filter out items ?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h (hunk 1 above)
Comment (uclaros) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:78]: `QWidget` has a `setEnabled` non virtual slot and I didn't want to shadow it / cause any confusion

Comment (troopa81) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:78]: Right!

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h [hunk 2]
```diff
@@ -16,17 +16,98 @@
...
-class QgsValueRelationWidgetFactory;
+///@cond PRIVATE
+
+/**
+ * \brief The QgsFilteredTableWidget class
+ *
+ * This is a helper widget for QgsValueRelationWidgetWrapper
+ * This widget is a QTableWidget showing checkable items, with an optional QgsFilterLineEdit on top that allows filtering the table's items
+ */
+class QgsFilteredTableWidget : public QWidget
```
[the remark this answers is not in the document]
Comment (uclaros) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:40]: The thing is that some of its members are used in the tests: `mTableWidget->item()`, `mTableWidget->rowCount()` and `mTableWidget->mSearchWidget->setText()`.
I would need to create some (private) methods in `QgsValueRelationWidgetWrapper` to handle those, which is a bit ugly. What do you think?

**Comment by uclaros** [comment]
> Have you considered using QAbstractItemModel instead of a QTableWidget since all the model logic is within mCache ? It will avoid allocating QTableWidget every time filter string changes, no?

I though about it a bit at first, but I found confusing the part that the model (which can be filtered) should be one dimensional and should be displayed in a two dimensional QTableView. It would probably need some extra intermediate proxy model to take care of the distribution of items to columns... so I went with reusing the simpler existing QTableWidget/QTableWidgetItem
I also though of adding a singleshot timer delay before applying the filter string to limit the table rebuilds between keystrokes, but did not see a need for it while testing with thousands of values.

> Besides that, is it possible to make the space between lines constant so it doesn't grow when we filter out items ?

Styling (and the rest of the functionality) was left unchanged, so this is how the table also looks when the total number of items is small.

A potential future improvement would be a context menu to offer check/uncheck all/filtered

**Comment by troopa81** [comment]
> I though about it a bit at first, but I found confusing the part that the model (which can be filtered) should be one dimensional and should be displayed in a two dimensional QTableView. It would probably need some extra intermediate proxy model to take care of the distribution of items to columns

I was more thinking about a 2-dimensionnal model taking care about the filtering, (with *data()* method containing quite the same logic of the *filterStringChanged* method), not a 1-dimensionnal model with proxysortfiltermodel which, I agree would be too complicated.

> Styling (and the rest of the functionality) was left unchanged, so this is how the table also looks when the total number of items is small.

OK

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (troopa81) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h:40]: I agree, I didn't think about the tests. No need to change anything

**Comment by uclaros** [comment]
> I was more thinking about a 2-dimensionnal model taking care about the filtering, (with _data()_ method containing quite the same logic of the _filterStringChanged_ method),

hmm, sorry, I'm still not sure how filtering would work without a filter proxy model that would also translate filtered indexes to original indexes.

**Comment by troopa81** [comment]
> hmm, sorry, I'm still not sure how filtering would work without a filter proxy model that would also translate filtered indexes to original indexes.

- one QAbstractItemModel
- data() returning elements according to the same logic that the one describe in filterStringChanged method,
- setData to modify mCache on check/uncheck
- rowCount/colCount where computation is based on the config parameters
- send signal layoutChanged() when the content of the filter line edit has changed 

And I think that's pretty it! 

QTableWidget is for display facilities and the Qt documentation state that 

> If you want a table that uses your own data model you should use [QTableView](https://doc.qt.io/qt-6/qtableview.html) rather than this class.

And I think that's kind of the situation here.

**Comment by uclaros** [comment]
> * rowCount/colCount where computation is based on the config parameters

rowCount is variable and you do not know it until all of mCache is checked against the filter string

> * data() returning elements according to the same logic that the one describe in filterStringChanged method,

filtering logic cannot live within the data() method. The accepted index range depends on the filtered results, hence the need for a proxy model.

**Comment by troopa81** [comment]
@uclaros Yes, you're right. Thanks for clarification!

I'm OK to merge, I'm restarting the CI because I don't get why the build artifact is missing

**Review by troopa81 — APPROVED** [review]

**Pull request merged**